src/host/services/IMUService.cpp

```diff
@@ -26,38 +26,204 @@
+      _signConfig = 0b00000000;
+  }
+
   DEBUG("Initing BNO055");
-  if (!bno055.begin()) {
+  if (!bno055.begin(bno055.OPERATION_MODE_NDOF, _axisConfig, _signConfig)) {
     DEBUG("Error initializing BNO055");
   }
   else {
-    DEBUG("Success!");
+    // read calibration and offset values for heel & pitch from flash
+    readCalibrationOffsets();
+    readHeelPitchOffset();
   }
 }
 
 void IMUService::loop() {
-  bno055.getCalibration(&sysCalib, &gyroCalib, &accelCalib, &magCalib);
+  bno055.getCalibration(&_sysCalib, &_gyroCalib, &_accelCalib, &_magCalib);
+  //DEBUG("Calib Sys: %i Accel: %i Gyro: %i Mag: %i", _sysCalib, _accelCalib, _gyroCalib, _magCalib);
 
   eulerAngles = bno055.getVector(Adafruit_BNO055::VECTOR_EULER);
-
-  //DEBUG("Calib Sys: %i Accel: %i Gyro: %i Mag: %i", sysCalib, accelCalib, gyroCalib, magCalib);
-  //DEBUG("Attitude roll: %f pitch: %f  Mag heading: %f", eulerAngles.z(), eulerAngles.y(), eulerAngles.x());
+  //DEBUG("Sensor Raw-datas: eulerAngles.z=%.3f eulerAngles.y=%.3f  eulerAngles.x=%.3f", eulerAngles.z(), eulerAngles.y(), eulerAngles.x());
 
   SKUpdateStatic<2> update;
   // Note: We could calculate yaw as the difference between the Magnetic
   // Heading and the Course Over Ground Magnetic.
 
-  /* if orientation == MOUNTED_ON_PORT_HULL */
-  double roll, pitch, heading;
-  roll = eulerAngles.z();
-  pitch = eulerAngles.y();
-  heading = fmod(eulerAngles.x() + 270, 360);
+  switch (_config.mounting) {
+    case VerticalPortHull:
+      _roll = SKDegToRad(eulerAngles.z());
+      _pitch = SKDegToRad(eulerAngles.y());
+      _heading = SKDegToRad(fmod(eulerAngles.x() + 270, 360));
+    break;
+    case VerticalStbHull:
+      _roll = SKDegToRad(eulerAngles.z())*(-1);
+      _pitch = SKDegToRad(eulerAngles.y())*(-1);
+      _heading = SKDegToRad(eulerAngles.x());
+    break;
+    case VerticalTopToBow:
+      _roll = SKDegToRad(eulerAngles.y());
+      _pitch = SKDegToRad(eulerAngles.z());
+      _heading = SKDegToRad(fmod(eulerAngles.x() + 180, 360));
+    break;
+    case HorizontalLeftSideToBow:
+      _roll = SKDegToRad(eulerAngles.y());
+      _pitch = SKDegToRad(eulerAngles.z()) * (-1);
+      _heading = SKDegToRad(eulerAngles.x());
+    break;
+  }
+
+  // update NavigationHeadingMagnetic if quality (= calibration value) is ok
+  if (isMagCalibrated()) {
+    update.setNavigationHeadingMagnetic(_heading);
+  }
+  // update NavigationAttitude if quality (= calibration value) is ok
+  if (isHeelAndPitchCalibrated()) {
+    update.setNavigationAttitude(SKTypeAttitude(/* roll */ _roll + _offsetRoll, /* pitch */ _pitch + _offsetPitch, /* yaw */ SKDoubleNAN));
+    //DEBUG("Heel = %.3f | Pitch = %.3f | Heading = %.3f", SKRadToDeg( _roll + _offsetRoll), SKRadToDeg( _pitch + _offsetPitch), SKRadToDeg(_heading));
+  }
+  _skHub.publish(update);
+
+  // Save calibration values to EEPROM every 30 Minutes, if BNO055 is fully calibrated
+  if (bno055.isFullyCalibrated() && _timeSinceLastCalSave > 1800000) {
```

Comment:
`_timeSinceLastCalSave` is never initialized. You should initialize it to the value of the timer so that it automatically saves the first time it's fully calibrated. And this should be a `static const` member of the class.

Reply:
First point is clear.

But why static const?
If the sensor looses calibration and later on will get calibrated again, the new values should be written again to flash I think.